Points and Pixels as Page Layout Units

sidewayss opened the pull request:
MuseScore Feature Request: 106861
here: https://musescore.org/en/node/106861

Jojo-Schmitz commented on libmscore/mscore.h:
```diff
@@ -68,11 +68,29 @@ inline int trackZeroVoice(int track) { return track & ~3;    }
+// 5 = 100% scaling, 6 = 120% scaling, 4.5 = 90% scaling, etc.
+//
+enum class Units : unsigned int {
+    PX = 0, // Pixels/Points
```
`unsigned char` would be enough here, wouldn't it? Actually probably even `char`, as here we don't really care whether signed or not?

Jojo-Schmitz commented on mscore/pagesettings.h:
```diff
@@ -36,15 +38,23 @@ class Navigator;
       Q_OBJECT
 
+      static const int PREVIEW_WIDTH  = 0; // This is really NON_HEIGHT, but WIDTH makes it a nice pair
+      static const int PREVIEW_HEIGHT = 1;
```
could be an `enum` too?

sidewayss replied:
Could be.  If it was 3 items I probably would have made it so.  2 is borderline.

Jojo-Schmitz commented on mscore/textprop.cpp:
```diff
@@ -96,7 +96,8 @@ void TextProp::setScore(bool onlyStyle, Score* score)
 void TextProp::mmToggled(bool val)
       {
-      QString unit(val ? tr("mm", "millimeter unit") : tr("sp", "spatium unit"));
+    QString unit(val ? tr(unitSuffixes[(int)Units::MM], "millimeter unit")
```
indentation...

Jojo-Schmitz commented on mscore/pagesettings.cpp:
```diff
@@ -631,14 +541,48 @@ void PageSettings::updatePreview(int val)
+    default: // Units::PX
+        return 1; // No conversion
+        break;
+    }
```
indentation

Jojo-Schmitz commented on libmscore/mscore.h:
```diff
@@ -68,11 +68,29 @@ inline int trackZeroVoice(int track) { return track & ~3;    }
+// This constant also doubles as the base scaling factor for a score:
+// 5 = 100% scaling, 6 = 120% scaling, 4.5 = 90% scaling, etc.
+//
+enum class Units : unsigned int {
```
`unsigned char` would be enough here, wouldn't it? Actually probably even `char`, as here we don't really care whether signed or not? Later you cast to `(int)` anyway, so the unsigned isn't used at all.

sidewayss replied:
Yes it could be char.

Jojo-Schmitz replied:
I'd use `int(Units:MM)`, looks more C++ like, less C like.
And no, making it an enum: char would not result it being interpreted as ASCII, when cast to int, a char is after all just an integer with a smaller range. As long as it is not used in a printf %c format specifier, it would not be confused for an ASCII sign. And it is not running from 1 to N, but from 0 to N-1 here, anyway It'd drop the unsigned. unless we want a -1 to mean INVALID or some such.

sidewayss replied:
OK, I see, so I'd use ascii 0, not '0'.  duh!  I can do that.  More commits to squash...

Jojo-Schmitz commented on libmscore/mscore.h:
```diff
@@ -68,11 +68,29 @@ inline int trackZeroVoice(int track) { return track & ~3;    }
+    PX = 0, // Pixels/Points
+    MM,     // Millimeters
+    INCH,   // Inches
+    SP      // Staff Spaces
```
indentation

sidewayss replied:
easily fixed - btw I've fixed existing indentation in various places several of the files in this PR.

sidewayss commented:
This code generates a lot of "unused" warnings because the constant array <code>unitSuffixes</code> is in mscore.h.  If that constant were to move to page.h, or pagesettings.h, then textprop.cpp (and maybe other files) were to include page.h or pagesettings.h, that could be eliminated.  See my attached doc for a listing of specific edits and some other questions/concerns.
[MuseScore_pt-px.pdf](https://github.com/musescore/MuseScore/files/238654/MuseScore_pt-px.pdf)

sidewayss commented:
The Travis CI build fails due to the file format change, the addition of the "page-units" tag.

sidewayss commented:
I'll wait to commit changes to indentation and other non-critical changes until the critical part of the code is deemed OK.  I don't want to clog this up with a lot of commits.

Jojo-Schmitz commented:
You could fix the issues pointed out so far and squash evrything into one commit to avoid the clogging. Eventually that'd be needed anyway ;-)

sidewayss commented:
I have now:
1) fixed the indentation issues, plus a couple you didn't catch :-)

2) eliminated all but 1 of the "defined but not used" warnings, by moving the Units definition to page.h and the unitSuffixes definition to pagesettings.h, then including pagesettings.h in textprop.h.
<i>I get 1 warning here</i>: \MuseScore\build.qtc\mscore\mscore_automoc.cpp:73
I am not familiar with that file.  What would be a plausible solution to eliminating this warning?

I am about to squash the commits, but I wanted to sync with github first, as I want that stored as my git reset --hard baseline.  Sorry about the extra churn here.  I'll commit the squashed commits when I have it ready.

I didn't change unsigned int to char, but I explained why in my response to Jojo's comment
I didn't make those PREVIEW constants an enum, but again, explained why in my comment

What is required to change the test(s) in the Travis build?  I assume it is something I am not even authorized to do, much less qualified.  But you tell me.

lasconic commented on libmscore/page.cpp:
```diff
@@ -399,9 +407,11 @@ void PageFormat::read(XmlReader& e, Score* score)
             else if (tag == "page-width")
-                  _size.rwidth() = e.readDouble() * .5 / PPI;
+                  _size.rwidth() = e.readDouble() / SCALE_XML;
+            else if (tag == "page-units")
```
I'm not sure it's a good idea to store the unit this way. Maybe it's better to save a string representing the unit. @wschweer Any opinion?
Also, we should pick a default, probably the unit used in MuseScore 2 and not store `page-units` at all if the default is used.

sidewayss replied:
If you want to use a string, the obvious candidate is unitSuffixes[] itself, with "mm", "in", etc..
Not storing the page-units if the value is Units::MM or "mm" is a very easy adaptation.  I assume you still want millimeters as the default.

Note that regarding default settings for the Page Settings dialog, there was a comment about making the units a global user setting in the code here:
https://github.com/musescore/MuseScore/pull/2573/files#diff-b8d24849ea6144a41df80b460106035aL50
And Marc suggested the same thing here, and my response below it (#8) too:
https://musescore.org/en/node/108196#comment-488171

A totally separate future PR, if it is to be done, but worth noting in this context.

sidewayss replied:
I also responded inline in the PR. Your email confused me at first 
because it leaves out the most important line of code, the one below 
your comment, the one immediately below the code you quote here in the 
email:

+ _units = (Ms::Units)e.readInt();

That is the line of code that reads the value as an int (the code you 
are quoting is the read code, not the write code, just to be clear).  
Here is my inline response to your comment:
https://github.com/musescore/MuseScore/pull/2573/files#r61304759

--
Sideways

On 4/27/2016 11:29 AM, Nicolas Froment wrote:
>
> In libmscore/page.cpp 
> <https://github.com/musescore/MuseScore/pull/2573#discussion_r61300930>:
>
> >              else if (tag == "page-width")
> > -                  _size.rwidth() = e.readDouble() * .5 / PPI;
> > +                  _size.rwidth() = e.readDouble() / SCALE_XML;
> > +            else if (tag == "page-units")
>
> I'm not sure it's a good idea to store the unit this way. Maybe it's 
> better to save a string representing the unit. @wschweer 
> <https://github.com/wschweer> Any opinion?
> Also, we should pick a default, probably the unit used in MuseScore 2 
> and not store |page-units| at all if the default is used.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/musescore/MuseScore/pull/2573/files/26336f34193fea4f4e4e6ad052c5624a411478b6#r61300930>
>

wschweer replied:
I dont see the need to change the current file format at all. The numbers in the file are intern and could stay at "mm". Simple convert them on load/save.

sidewayss replied:
The file format is only changing via the addition of the "page-units" tag.  I believe the question here is about the contents of the "page-units" tag being an integer, Units, or a string, unitSuffixes[].  Do we store the number one, or the string "mm", for example?
The units used in the file storage for page settings values have not changed.  They have always been at 144dpi.

wschweer replied:
ok, i was misreading the code.
My preference would be storing a string.

sidewayss replied:
OK, I'll make that change.  I was also confused by the original line note, the selected line should probably have been one line down, the one below this note.
I can see how storing the string is more human-readable within the file, more xml-like.
This is entangled with the unitSuffixes[] discussion elsewhere, and I'm still wondering about the translations, and if they need to be integrated into the Page Settings dialog, and if that has any ripple effects here.  I'll get this going and see how that other discussion evolves...

sidewayss commented:
I just squashed my commits locally, and everything looks good.  Then I realized there had been an update to the master, so I fetched, rebased, and now the commit history has the Merge Branch as a commit.  Is that a problem (I assume it is)?  git rebase -i isn't letting me squash it, as it doesn't even appear in the file.  Any advice?

lasconic commented:
You must do it in the wrong order.
you need to do
```
git checkout master
git fetch upstream
git merge upstream/master
git checkout yourbranch
git rebase master
git push -f origin yourbranch
``` 
as explained in the [git workflow](https://musescore.org/en/developers-handbook/git-workflow#Rebase-your-branch-on-the-latest-upstream). No "merge branch" commit should appear since you don't merge any branch

sidewayss commented:
Latest changes address the char vs. unsigned int discussion, and also implement the "do not save the page-units if they are the default value (mm)" code.  I'm waiting for the Travis build to complete (successfully!?).

After various attempts and some reading, I have now also successfully squashed everything.  I will push that once this Travis build is complete.

sidewayss commented:
I have just pushed the squashed stuff and everything looks ok on githum.com and my local machine.
The Travis build is failing on something relating to clefs, nothing to do with my code, apparently.

sidewayss commented on mscore/pagesettings.ui:
```diff
@@ -94,7 +94,7 @@
           </property>
           <property name="maximum">
-           <double>2000.000000000000000</double>
+           <double>200000.000000000000000</double>
```
Jojo - do you remember the discussion we had late last year about rounding errors in Page Settings?  You had suggested limiting the values in the dialog box to 3 digits, since that's what it is elsewhere in MuseScore.  I think it was 3 digits...  Now would be a good time to do that, right?  And it's simply a change in the settings in the Qt dialog editor UI, right?
Here is where a rather lengthy discussion begins: https://musescore.org/en/node/61746#comment-388036

Jojo-Schmitz replied:
rouning to 3 Digits sounds resonable to me.

sidewayss commented:
FYI: This PR fixes these issues directly:
https://musescore.org/en/node/91811
https://musescore.org/en/node/108196

It fixes these coincidentally:
https://musescore.org/en/node/108226
https://musescore.org/en/node/97291

sidewayss commented:
A concern of mine, mentioned in the .pdf attached above:
scoreview.cpp, line 1858:
https://github.com/musescore/MuseScore/blob/master/mscore/scoreview.cpp#L1858
I did not make any changes to this file, but I'm wondering if that's a mistake.  This line of code calls the modified margin functions in page.cpp.  Those functions no longer multiply by DPI.  That should break this scoreview.cpp code, but it doesn't seem to, or I don't know where to look.

Jojo-Schmitz commented:
To me the Travis failure does look related, it is about page format. Nothing to with clefs, it is just a clefs test files that shows strange behavoir with ist page format.

sidewayss commented:
OK, can  you explain it to me?  I don't understand how to debug this.  Thanks.

Jojo-Schmitz commented:
`4468-          <right-margin>1.06581e-13</right-margin>`
`4469+          <right-margin>1.9984e-14</right-margin>`

Looks like a rounding problem to me, both valuea are very small

sidewayss commented:
I found my local copy of the file, but not the reference file.
Interesting right after I brought up rounding...
How to resolve this?  I just opened my local version of the test file, clef-1-mscx.  The even right margin is zero.  Time to add the rounding change I just brought up?  Or will that still cause this conflict with the reference version of clef-1-mscx?

sidewayss commented:
OK, i get it.  I just opened the local clef-1 file, then saved it as uncompressed, and compared against the original.  This is the exact difference.  Note that the odd margins are all zero, and the even margins are generally derived from the pageWidth - leftMargin - visibleWidth.  I have reset the even right margin to 0 manually, resaved and this is the same.
If I attempt to manually enter 210mm as the page width, the page layout goes crazy!  This is a custom page size and there is something funny going on here.  Try re-entering 210 in the page width and see what I mean.

Jojo-Schmitz commented:
there is clef1-ref.mscx too. Opened in MuseScore it does Show 0 margin, but in a text Editor it does show `<right-margin>1.9984e-14</right-margin>`

Jojo-Schmitz commented:
there is clef1-ref.mscx too. Opened in MuseScore it does Show 0 margin, but in a text Editor it does show `<right-margin>1.9984e-14</right-margin>`, so yes, rounding now seems to be a good time

Jojo-Schmitz commented:
there is clef1-ref.mscx too. Opened in MuseScore it does ahow 0 margin, but in a text Editor it does show `<right-margin>1.9984e-14</right-margin>`, so yes, rounding now seems to be a good time.
Changing the file(s) to have a 5mm margin should appease Travis too I guess

sidewayss commented:
Yea, I was just playing with clef-1-ref, and it behaves better than clef-1 when it comes to changing the page width value. If I change it to 210, the same value it shows, then save the file as..., it changes the file value to a larger number (still small), but representable without scientific notation: <code>0.0011811</code>
But we'll need to change the version that Travis uses too, right? I'll start looking at the necessary rounding changes.

sidewayss commented:
FYI: the original discussion about rounding, above, was only about the Staff Space field, the spatium value.  This here is something else.
1) Before and after this PR, millimeters are converted units, converted by multiplying.  That multiplication of the value in the dialog causes lots of digits after the decimal in the result.
2) Then the subtraction begins.  Right margins are not stored, they are calculated:
<code>pageWidth - printableWidth - leftMargin</code>
So even if the numbers in millimeters calculate a clean right margin of  zero millimeters, that same subtraction on the multiplied values may not.  Though the difference I'm looking at right now in the debugger starts at the 4th decimal place, which means a) yes, rounding to 2 or 3 digits will solve the problem, but b) why would multiplying 3 values by a factor, then adding them back together be so far off when the math is done at full precision?  I would expect to see values more like the scientific notation in the clef-1 files.
3) So this means rounding the converted values, stored in memory at 72dpi, and in the files at 144dpi.
4) This does not just affect intended zero values as in this case.  This will likely require opening and resetting a bunch of test files' Page Settings.

sidewayss commented:
I did the math for the clef-1 files' even right margin in an Excel spreadsheet that I have attached. If you don't have excel, I have attached a .pdf, but you can't see the formulas in the pdf file.  I wanted to have an outside reference for the larger numbers I was seeing in the debugger.

After manually re-inputting the margins and page width, the even right margin value I'm seeing in the file is exactly the same as the calculated number in Excel: -8.17124e-14 vs. -8.17124E-14.  Even the same representation in scientific notation. That's good news.

I'm concerned about rounding these numbers, but I'll give it a try and see how it goes.  But I think the main problem for the Travis build is that these older files need resetting because the previous conversions used inches as an intermediary unit, and that messes with the unrounded values.  I would expect to see more of this in other files.
[MsPsRounding.xlsx](https://github.com/musescore/MuseScore/files/239457/MsPsRounding.xlsx)
[MsPsRounding.pdf](https://github.com/musescore/MuseScore/files/239459/MsPsRounding.pdf)

sidewayss commented:
Turns out that rounding floating point numbers to 3 decimal places in C++ is not a reliable thing.  It works really well when you're converting to a string, but the type <code>double</code> can't store 595.276, the width of the clef-1 file's page converted to points, for example.  It stores 595.275999999.  Nothing you can do about it.

Apparently there are libraries that implement a "decimal" data type that deals with this stuff, but none in Qt.  I'm too used to Excel, where the basic ROUND() function takes a second argument indicating the number of decimal places.

The values can be rounded to 3 digits by converting to strings prior to saving to the xml.  That might be a good idea.  It's the only place it's plausible.  And it will cause more test files to require resetting.

Jojo-Schmitz commented:
Floating point arithmatic really sucks... But storing a 0 as -8.17124e-14 is pretty insane, we could detect 'near 0' and make that a real 0.
Then again right margins are stored, so it should be the printable area that gets calculated

Jojo-Schmitz commented:
To just get that one particular mtest pass, take the clef-1.mscx and clef-1-ref.mscx into a plain text editor and replace the near 0 value for right margine with the value used for left margin.

sidewayss commented:
I also considered the "near 0" rule.  It will eliminate the scientific notation.  I'll go that route and see if we hit any other files with the issue.

The printableWidth is also calculated, but only once, in the Page Settings dialog, whenever you set the left margin or page width.  But then it's stored in the PageFormat object.  It's a bit goofy, but I'm hesitant to change that whole mechanism.

sidewayss commented:
Hard to believe it, but the clef-1 files were the only ones with this issue.  The Travis build passed!

Jojo-Schmitz commented:
presumably it is the only file with a 0 margin.
You're now rounding less then 3 digits, in the case of 0. I'd use .0005 not .005 here, that way a poossibly lated implemented rounding at 3 digits won't conflict?

sidewayss commented:
I went with 2 digits because the dialog can only represent 0.01 as the smallest value.  Though you're correct that It's converted to points, so it could use more precision.  Though a margin smaller than 0.01 points is not exactly visible on the page.
But it doesn't really matter much, and my formatting on those lines of code was funky anyway, so I''ve commited a change to 0.0005.

sidewayss commented:
I just squashed the recent commits, and in doing so realized how those merges where showing up in the commit history: If I go back to far in the commit history with the git rebase -i command, it starts including those master commits in the history.  Sometimes it difficult to know what number to put in the HEAD~N part of the git rebase -i, and I go too far back.

Jojo-Schmitz commented:
You can still do a `git commit --amend`, fix the commit message and then do a `git push -f`

Jojo-Schmitz commented:
I believe if you'd use
`#ifdef Q_MOC_OUTPUT_REVISION`
`static const char* unitSuffixes[] = {"px", "mm", "in", "sp"};`
`#endif`
this would fix the last remaining compiler warning, the one about it being unused?

wschweer commented on libmscore/page.cpp:
```diff
@@ -292,15 +293,21 @@ void Page::scanElements(void* data, void (*func)(void*, Element*), bool all)
+      _evenLeftMargin   = marginPX;
+      _oddLeftMargin    = marginPX;
+      _printableWidth   = _size.width() - marginPX2;
+// ??? Why no printableHeight???
```
Printable width for even/odd pages must be the same to allow for faster/easier layout code.
Printable height can be different, so top and bottom margins for even/odd pages can be independent.

sidewayss replied:
Would you like me to include that info as a comment here or simply delete my comment question?

wschweer replied:
If you want you can move it to the PageFormat header...

sidewayss replied:
I just did that, and eliminated the comment about margins being stored in inches in page.h, as that is no longer the case.  It will be in my next commit.

Jojo-Schmitz commented:
or you just declare it in pagesettings.h (`static const char* unitSuffixes[];`)and define and initialize it in pagesetting.cpp. This should be the better fix anyhow?
No code generating stuff belongs into .h files after all.

wschweer commented on mscore/pagesettings.h:
```diff
@@ -22,29 +22,42 @@
 class Navigator;
 
+// For converting from Ms::Units to a two-char string used as a suffix
+static const char* unitSuffixes[] = {"px", "mm", "in", "sp"};
```
this line should not be in a header file; what about translations?

sidewayss replied:
It is only used in one place that is not a debug message or internal to the PageSettings dialog box (pagesettings.cpp).  And that one place is here:
https://github.com/musescore/MuseScore/pull/2573/files/831df406b6bd81c4e3a2f79c4551410470f343c0#diff-53fdc139c4fd1972f05706a81c9ee7c0R100

Jojo-Schmitz replied:
I basically said the same as to where it belongs, also to fix that compiler warning this causes. As to translations, I don't think we should translate them, these are units, and as such international?
But then again we do translate mm, sp and inch elsewhere (I've just checked to be sure), so indeed should here too. Guess a bunch of `Q_TRANSLATE_NOOP()` should do here?

sidewayss replied:
Are there additional places in the code where it could be used?  I only searched for the literal values of "mm" and "in" and "sp", etc. Looks like I missed these in inspectorImage.cpp here, but that's all I see:
https://github.com/musescore/MuseScore/blob/master/mscore/inspector/inspectorImage.cpp#L180-L212

Jojo-Schmitz replied:
thinink about it some more: that are quite a few places we we use unit suffixes (and visible in the UI and some translated), in addition to the ones here we use "%", "" (i.e. an empty string) and "pt" (for font size in Point). Maybe we could once and forever collect and combine them in just one spot? Also only translate them once...

sidewayss replied:
Could you elaborate so that I might investigate?  I'm unfamiliar with MuseScore's translation code, and I don't know what to search for in the files to see these usages of "pt" or "%".
The points vs. pixels situation is interesting, in that they are equivalent.  I combined them, but they could be separated in unitSuffixes and Units - though that depends on whether you want to separate them on the Page Settings dialog, which is a whole other deal.  Reviewing the places "pt" is used will illuminate all of that.

Jojo-Schmitz replied:
just search for setSuffix. Inyout arrax use something like ` QT_TRANSLATE_NOOP3("suffixUnits", "mm", "millimeter unit")` and in the cose rather than the litetal `"mm"` use `qApp()->translate("suffixUnits", suffixUnits[1])`, mabe instead of the digit and an enum, for better readability, `int(SuffixUnit::MM)`.

sidewayss replied:
OK.  What about the literals I pointed out elsewhere in inspectorimage.cpp?  Are those exposed to the user and they should be translated too?  See here:
https://github.com/musescore/MuseScore/blob/master/mscore/inspector/inspectorImage.cpp#L180-L212

Jojo-Schmitz replied:
yes, they are

sidewayss replied:
I'm seeing a figuredbass.h setSuffix() too.  Seems like I should exclude calls to that function from these changes, correct?

Jojo-Schmitz replied:
Yes, ignote that one from fifured bass, it gets it directly from xmlimport.
"inch unit", just like it is in one plase already should be fine. It is the hint to translators, so than know ehaz "in"  and don't confuse it for the opposite of "out".
"pt" is used in mscore/editstafftype.ui, so not in a .cpp file, but is a suffix troo and could get overwritten with  setSuffix(), same for sp in sevaral places of several .ui files. I see also "ms"(miliseconds), "dB" (deziBel), "BPM" (Beats Per Minute), "sec" (Seconds) and "Hz" (Hertz) in .ui files

sidewayss replied:
OK.  As for milliseconds, decibels, bpm, etc - those are not at all related to PageFormat.  Do you really need to include those in this PR?  That's a bunch more files to touch.  It seems that the % and empty string might be outside that scope as well, since they appear to be unique to drumroll.cpp and pianoroll.cpp (even though "roll" means something entirely different in those file names).

Jojo-Schmitz replied:
Hmm, guess you're right it'd be too much. And yes the QT_TRANSLATE_NOOP3 is for the extra 3rd arg, the translation hint

sidewayss replied:
I'm compiling my first pass of this now.  It doesn't cover the font size in points in editstafftype.cpp, but it will before I commit it.  It will be setup to expand into time units (sec, ms, bpm) and audio units(db, hz).

sidewayss replied:
I'm have a difficult time separating the declaration of unitSuffixes from the definition/initialization, mostly because it is entirely outside of any class, and because it involves strings.  I have tried a variety of things, including what I consider to be the ideal, which is declaring unitSuffixes this way:
<code>std::map&lt;Ms::Units, char*&gt;</code>
or like this:
<code>QMap&lt;Ms::Units, QString&gt;</code>
But without a class constructor I don't know where to initialize it.  And this doesn't compile:
<code>QString(QT_TRANSLATE_NOOP("context", "value");</code>
QString() doesn't like the curly braces inside the macro.
My lack of computer science education is exposing itself here.

Jojo-Schmitz replied:
_without-name = "%" (i don't know what to call it)_  how about "Percent"? 
 `static const char* unitSuffixes[];`in  and something like this in the corresponding cpp file:
 `unitSuffixes[] = {QT_TRANSLATE_NOOP3("unitSufixes", "px", "pixel units), QT_TRANSLATE_NOOP3("unitSuffixes", "mm", "millimeter unit"), ....};`

sidewayss replied:
I tried that, and a bunch of variations on it, and the compiler choked every time except once.  And that one version didn't store the strings, it was all empty, so the memory wasn't being allocated, or something...
Is there a global, single instance of some class that I can use to stash this?  I could make it a property of that class.

Jojo-Schmitz commented on mscore/pagesettings.cpp:
```diff
@@ -138,24 +133,37 @@ void PageSettings::blockSignals(bool block)
+            break;
+      }
+
+      pageWidth->setSuffix(suffix);
```
I think here an untranslated suffix would be exposed to the user. But if so before your change it was too

sidewayss replied:
Yes, none of the suffixes displayed in the Page Settings dialog are translated, and never were AFAIK.

sidewayss commented:
<i>to be clear on the printableWidth discusion above:</i>
Jojo could be correct, that the file should store printableWidth instead of right margins, or that the code should store right margins instead of printableWidth.  The code is out-of-sync with the file in this respect.  But I think that such a change would cause backwards compatibility problems, and cause changes in this code that would demand a bunch of further testing.  It's something that can be tackled at a future date if desired.

sidewayss commented:
The latest commit adds translation, but doesn't move the definition of unitSuffixes, as that is causing me issues.  So the 1 warning is still present as well.

sidewayss commented:
OMG, I got it to work.  I was sure I had tried this exact same setup previously, but I was doing the NOOP3 mixed with changing the definition, and that must have been my issue.  I don't know.  It is committed, it saves the units in the file as an untranslated string.  I think that's all the outstanding issues mentioned.

sidewayss commented:
I have now added translation to the suffixes in inspectorImage.cpp.  I added them to editstafftype.cpp too, but then encountered the existing issue (opening this dialog crashes MuseScore):
https://musescore.org/en/node/108146

I have updated that issue with information from my debugging session.
The latter required the addition of a new Unit, Ms::Units::PT, for points only, for font sizes.  I left that in there for when issue 108146 is resolved.  But I didn't want to make changes to that file before resolving that issue.
Units and unitSuffixes are in page.h/page.cpp.  So using them requires including page.h, which none of these files do now  If this is to evolve into audio and time units (milliseconds, hertz, etc.), page.h might not be the place for these. Is that the direction this is headed, and if so, in which file do they belong?

sidewayss commented on mscore/pagesettings.cpp:
```diff
@@ -189,65 +197,47 @@ void PageSettings::updateValues()
-      if (twosided->isChecked()) {
-            evenPageRightMargin->setValue(oddPageLeftMargin->value());
-            evenPageLeftMargin->setValue(oddPageRightMargin->value());
+      // ???Why no setMaximum for height???
```
The code above this line note sets max values for the left and right margins.  There is no code to set max values for the top or bottom margins.  Is there a reason for that, or should it be added?

sidewayss commented on mscore/pagesettings.cpp:
```diff
@@ -130,6 +124,7 @@ void PageSettings::blockSignals(bool block)
       pageHeight->blockSignals(block);
       pageOffsetEntry->blockSignals(block);
       pageGroup->blockSignals(block);
+// ???Radio button signals not blocked???
```
The current code never blocks signals for the radio buttons.  Is this by design, or should I add it?  I haven't seen any problems, but I haven't stress-tested the UI either.

sidewayss commented:
I have 2 questions about this line of code in harmonyedit.cpp:
https://github.com/musescore/MuseScore/blob/master/mscore/harmonyedit.cpp#L161

1) Should I remove the <code> * DPI / PPI</code>?  DPI/PPI = 1, so it doesn't do anything
Note that this is the only place in the code that currently uses the constant PPI
2) What is the literal 20.0 representing here?  Is looks suspiciously related to SPATIUM20, and I'm wondering if there's a way to replace the literal with an existing constant or combination thereof.

MarcSabatella commented:
That file (harmonyedit.cpp) is implementing an experiment chord symbol
style editor that is currently disabled by default. My guess is that the
lines you are looking at are holdovers from before the big DPI change and
this file was overlooked since it isn't really relevant right now.  The
code kind of works, but would need a lot of attention to really be able to
use this.  Definitely something I'd like to follow up on when I have time.

On Fri, Apr 29, 2016 at 2:45 PM sidewayss <notifications@github.com> wrote:

> I have 2 questions about this line of code in harmonyedit.cpp:
>
> https://github.com/musescore/MuseScore/blob/master/mscore/harmonyedit.cpp#L161
>
> 1) Should I remove the * DPI / PPI? DPI/PPI = 1, so it doesn't do anything
> Note that this is the only place in the code that currently uses the
> constant PPI
> 2) What is the literal 20.0 representing here? Is looks suspiciously
> related to SPATIUM20, and I'm wondering if there's a way to replace the
> literal with an existing constant or combination thereof.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/musescore/MuseScore/pull/2573#issuecomment-215877159>
>

sidewayss commented:
Now I'm wondering the various conversions applied to the MStyle's  spatium value here:
https://github.com/musescore/MuseScore/blob/master/mscore/preferences.cpp#L294
here:
https://github.com/musescore/MuseScore/blob/master/libmscore/style.cpp#L683
and here:
https://github.com/musescore/MuseScore/blob/master/libmscore/style.cpp#L825

Just wondering if these conversions are necessary.  They are not directly related to this PR, but clearly related.  I want to make sure I'm not missing anything.

sidewayss commented on mscore/musescore.cpp:
```diff
@@ -4945,11 +4945,11 @@ int main(int argc, char* av[])
             PageFormat pf;
             pf.setSize(&ps);
-            pf.setPrintableWidth(psf.width() - 20.0 / INCH);
+            pf.setPrintableWidth(psf.width() - 20.0 / MMPI);
```
What does the 20.0 represent? Staff height?

sidewayss commented:
https://github.com/musescore/MuseScore/blob/master/libmscore/style.cpp#L825
One more concern: This PR makes getTenthsFromInches() obsolete.  These lines of code still call it, but they are inside a big block of disabled code.  What is the plan for this block of disabled code?  Should I be modifying these lines in anticipation of them being reinstated?

sidewayss commented:
fyi - As soon as #2578 is merged, I can test my changes to editstafftype.cpp for the translations and unitSuffixes there.  editstafftype.cpp is already including page.h somewhere in its include chain, so I didn't need to include is explicitly.  These changes includes the addition of the 5th unit, Points (suffix "pt").  Points are at least marginally related to Page Format, and this is a first test of expanding the scope of these units and their suffixes.
The changes are tiny, so the testing should quick and easy.  After that, adding %, empty string, bpm, hz, etc. will be straightforward.  But at that point, these will definitely be more than Page Format Units, so they should probably be somewhere other than page.h.

sidewayss commented:
I was able to rebase this normally until today, which is, of course, the day that I want to push changes because #2578 is merged and this code touches editstafftype.cpp.
What is the best approach/strategy to rebase this branch?  It is telling me that page.cpp, then layout.cpp is where it's choking, but the changes to those files since yesterday, especially layout.cpp are insignificant.  I even updated my branch manually with the page.cpp changes, and it reports the same error:<code>
CONFLICT (content): Merge conflict in libmscore/page.cpp
Auto-merging libmscore/layout.cpp
error: Failed to merge the changes.
</code>

sidewayss commented:
I was unable to rebase this branch properly, so I have created a new one that is now up-to-date with the master.  I have opened a new PR: #2624, but I am leaving this PR open so that the comments don't disappear.  If that's not an issue, then I'll close this PR and we can move to the new one.

lasconic commented:
The PR will still be around and it's linked in the other issue. I close this one.
